feat(#8224): add config to disable init endpoints

tomflenner opened the pull request:
# Description

This PR introduces two new `dapr run` arguments and corresponding `dapr.io` sidecar annotations to disable the automatic HTTP calls to `/dapr/subscribe` and `/dapr/config`. These changes are designed to reduce unwanted log noise and prevent edge case failures when an application does not use programmatic subscriptions or config loading. 

## Issue reference

Please reference the issue this PR will close: #8224 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [ ] Extended the documentation / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_
- [ ] Specification has been updated / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_ 
- [ ] Provided sample for the feature / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_

tomflenner committed:
feat(8224): add disable-init-endpoints argument

Signed-off-by: tomflenner <flennertom@gmail.com>

tomflenner committed:
feat(8224): disable /subscribe endpoint

Signed-off-by: tomflenner <flennertom@gmail.com>

tomflenner committed:
feat(8224): disable /config endpoint

Signed-off-by: tomflenner <flennertom@gmail.com>

tomflenner committed:
feat(8224): refactor /dapr/subscribe and test

Signed-off-by: tomflenner <flennertom@gmail.com>

JoshVanL reviewed (changes requested):
Please also include [integration tests](https://github.com/dapr/dapr/tree/master/tests/integration) for this change.

tomflenner committed:
feat(8224): switch to string array instead

See : https://github.com/dapr/dapr/pull/9077#discussion_r2384452610
Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
feat(8224): expose options as an annotation in injector code

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
test(8224): add integration test

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner commented on cmd/daprd/options/options.go:
```diff
@@ -167,6 +168,7 @@ func New(origArgs []string) (*Options, error) {
 	fs.IntVar(&opts.AppHealthProbeTimeout, "app-health-probe-timeout", int(config.AppHealthConfigDefaultProbeTimeout/time.Millisecond), "Timeout for app health probes in milliseconds")
 	fs.IntVar(&opts.AppHealthThreshold, "app-health-threshold", int(config.AppHealthConfigDefaultThreshold), "Number of consecutive failures for the app to be considered unhealthy")
 	fs.StringVar(&opts.AppChannelAddress, "app-channel-address", runtime.DefaultChannelAddress, "The network address the application listens on")
+	fs.StringVar(&opts.DisableInitEndpoints, "disable-init-endpoints", "", "List of initialization endpoints to disable. Supported values: config, subscribe,. If not specified, all endpoints will be enabled")
```
Done ✅

tomflenner commented on pkg/runtime/config.go:
```diff
@@ -455,6 +457,23 @@ func (c *Config) toInternal() (*internalConfig, error) {
 		return nil, fmt.Errorf("invalid value for 'app-protocol': %v", c.AppProtocol)
 	}
 
+	allowedEndpoints := map[string]struct{}{
```
Discussed with @JoshVanL few times ago on discord, i've moved const into `/pkg/runtime/config.go`  ✅

tomflenner committed:
test(8224): switch from string to array type

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
refactor(8224): golangci-lint array allocation

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
chore(8224): support extra ARGS for test-integration*

This allow the usage of the `-focus` argument in order to run specific test cases

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
refactor(8224): goimports

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
test(8224): improve subscription assertion

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
test(8224): http subscription integration tests

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
test(8224): http init config test

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner commented:
@JoshVanL i've added integration test for subscription and config but I am not sure adding one that cover both at same time is useful ?

Tell me if I am wrong here and if I need to add extra case, I thought about adding it under `startup/` test suite if required.

tomflenner committed:
chore(8224): goimports

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner committed:
chore(8224): typo after rebase

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

tomflenner commented on Makefile:
While we can run tests directly with `go test` (see https://github.com/dapr/dapr/blob/master/tests/docs/writing-integration-test.md#invoking-the-test), I've found it helpful to also support running via the makefile with `gotestsum`, which still allows using `-focus` arguments.

JoshVanL committed:
Merge branch 'master' into feat/8224-disable-init-endpoints

tomflenner committed:
chore(8224): typo on arg

Signed-off-by: backtrack5r3 <flennertom@gmail.com>

JoshVanL committed:
Merge branch 'master' into feat/8224-disable-init-endpoints

JoshVanL committed:
Merge branch 'master' into feat/8224-disable-init-endpoints

JoshVanL committed:
Merge branch 'master' into feat/8224-disable-init-endpoints

marcduiker commented:
@holopin-bot @tomflenner runtime-badge Thank you! Here's a digital badge as a small token of appreciation.
